Migrate to schnauzer v2 for configuration file templates

cbgbt opened the pull request:
**NOTE** This is against the `ootb-settings-extension` feature branch.
---

**Issue number:**
#3133

**Description of changes:**
* Adds support for "optional" settings extensions in templates. This lets us e.g. use AWS settings if the extension is installed, but not otherwise. Right now, the implementation does nothing because there's no concept of a setting being present or not.
* Uses the new schnauzer v2 rendering library in `thar-be-settings`, and changes all current templates to the new format
* Simplify schnauzer implementation by moving some `Option` types to have a `serde(default)`.

Existing templates were found by checking for all files installed into `%{_cross_templatedir}`.


**Testing done:**
* [x] All built-in test suites pass
* [x] Booted and checked logs/behavior of a k8s and ecs variant
* [x] Fuzz testing (I'm generating a corpus of settings objects and asserting that the old template system produces the same results as the new one)


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

cbgbt committed:
Migrate config file rendering to schnauzer v2

Bottlerocket invokes thar-be-settings whenever it needs to re-render
configuration file templates defined in the Bottlerocket API. This
change modifies thar-be-settings to use the schnauzer v2 renderer, and
also modifies all existing configuration templates in Bottlerocket to
use the new template format.

cbgbt committed:
schnauzer: allow optional extensions

cbgbt committed:
schnauzer: simplify template implementation

cbgbt commented:
This is what I get for doing some rebasing and not re-running the build 😅

bcressey commented on packages/containerd/containerd-config-toml_basic:
```diff
@@ -1,3 +1,4 @@
++++
```
I keep seeing this and wondering if we should make the `+++` delimiter optional, but then I think what a weird template this is and start wondering why it's a template at all.

We've gotten to a good place as far as ownership for the settings themselves and for related helper functions, but we still don't have a clear picture for who should own the template, which seems like it's inherently a cross-cutting concern that has to reflect the builder's intent.

For Kubernetes, we try to have a one-size-fits-all template, but it ends up requiring a lot of extensions. For containerd, we package all the templates with the software, which leads to this no-op template.

cbgbt replied:
I believe we must always require the `+++` delimiter at the start of a template, otherwise it is impossible for the parser to appropriately disambiguate a template body with `+++` in it, especially if that body is a TOML document.

As far as who owns the templates, I've been thinking they may want to live in their own sub-packages that can be overridden, and use e.g. `Provides: template(kubelet-kubeconfig))`

cbgbt replied:
Re-reading this while not being in a rush, I agree that this probably shouldn't be a template at all.

bcressey commented on packages/kubernetes-1.24/kubelet-exec-start-conf:
```diff
@@ -1,3 +1,7 @@
+[required-extensions]
+kubernetes = { version = "v1", helpers = ["any_enabled"] }
```
Is it possible to make `any_enabled` a `std` helper or does it assume too much about the data shape to work for anything other than Kubernetes?

cbgbt commented:
^ Rebases onto develop

cbgbt committed:
schnauzer: move any_enabled helper to std

cbgbt commented:
The above:
* Fixes two templates that changed in `develop` during the rebase
* Moves the `any_enabled` helper to `std` as requested by @bcressey

cbgbt committed:
schnauzer: fix helper composition bug

In order for helpers to be composable, they must implement the
`call_inner` method for the `HelperDef` trait. By failing to implement
this for `SettingExtensionTemplateHelper`, helpers used e.g. in an `if`
block would not return correct results.

cbgbt commented:
This should be ready to merge. I finished fuzz testing, which uncovered two bugs:
* I missed a comma in one of the template headers
* helpers in `handlebars` implement either `call` or `call_inner` depending on how they are defined. `call_inner` is composable, but `call` is not. I modified our wrapper helper to pass calls to both through to the wrapped helper.

cbgbt commented:
Some details about fuzz testing:
* I implemented random model generation for Bottlerocket in [this branch](https://github.com/cbgbt/bottlerocket/tree/randogen). It's a mess and can't really be done in a maintainable way to merge upstream, but it's helpful here.
* I generated [5k settings JSON blobs for each variant](https://github.com/cbgbt/bottlerocket-generated-models)
* I [wrote a tool](https://github.com/cbgbt/bottlerocket/blob/injectable-renderers/sources/api/schnauzer/src/bin/playdate.rs) which clones two Bottlerocket branches and renders a list of templates from both, one with `schnauzer::v1` and the other with `schnauzer::v2` and compares the ouput.
* The compared templates are [these](https://gist.github.com/cbgbt/e48be96b6c9c8c3c08f3c37b4ed62080)
* I did have to make a few changes to helpers to avoid network calls and to make iteration through `HashMaps` deterministic, but otherwise left things unchanged.


All 135,000 models now render identically for both tools.

stmcginnis reviewed (approved):
Great testing coverage!
